rope/contrib/generate.py

```diff
@@ -10,12 +12,40 @@
 )
 from rope.refactor import functionutils, importutils, sourceutils, suites
 
-
-def create_generate(kind, project, resource, offset, goal_resource=None):
+if TYPE_CHECKING:
+    from rope.base.pyobjects import PyObject
+    from rope.base.resources import Resource
+
+GenerateKind = Literal[
+    "class",
+    "function",
+    "module",
+    "package",
+    "variable",
+]
+
+GenerateVal = Union[
+    "_Generate",
+    "GenerateClass",
+    "GenerateFunction",
+    "GenerateModule",
+    "GeneratePackage",
+    "GenerateVariable",
+]
```

Comment:
This is accurate annotation, but I think this would be promising too much to the caller.

`create_generate()` is a factory method for creating refactoring operation; so rather than documenting the exact types that it can return, we should just document what it is able to promise: `create_generate()` will return an object that matches the Refactoring interface.

rope currently doesn't have a formal interface for Refactoring class, but if it does, it would be an `Protocol` or `abc` that looks somewhat like this:

```python
class Refactoring(typing.Protocol):
    def get_changes(self, *args, **kwargs) -> ChangeSet:
        ...
```

callers of `create_generate()` should only depend on it returning a class that matches the Refactoring interface, and they shouldn't rely on the concrete classes.

Reply:
Alternatively, we can make `create_generate()` return a more specific `GenerateRefactoring` Protocol, which can further add a promise that `get_changes()` doesn't take additional parameters:

```python
class GenerateRefactoring(typing.Protocol):
    def get_changes(self) -> ChangeSet:
        ...
```

this is more risky to promise in the API, since if we later change our mind and want to add a parameter to `_Generate.get_changes()`, that would need to be an API breaking change.